chain-time/src/timeframe.rs

```diff
@@ -0,0 +1,168 @@
+use crate::timeline::Timeline;
+use std::time::{Duration, SystemTime};
+
+/// Identify a slot in a *specific* timeframe
+///
+/// The slots are not comparable to others slots made on a
+/// different time frame
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Slot(pub(crate) u64);
+
+impl From<Slot> for u64 {
+    fn from(s: Slot) -> u64 {
+        s.0
+    }
+}
+
+/// Identify a slot in a specific timeframe and a leftover duration
+#[derive(Debug)]
+pub struct SlotAndDuration {
+    pub slot: Slot,
+    /// The offset of a specific time frame in
+    pub slot_duration_offset: Duration,
+}
+
+/// Time frame which is a timeline that is configured to be split in discrete slots
+pub struct TimeFrame {
+    timeline: Timeline,
+    slot_offset: Slot,
+    slot_duration: NbSeconds,
+    //slot_configuration: Vec<u64>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct NbSeconds(u64);
+
+impl NbSeconds {
+    pub fn new(seconds: u32) -> Self {
+        assert!(seconds < 600);
+        NbSeconds(seconds as u64)
+    }
+
+    pub fn to_duration(&self) -> Duration {
+        Duration::from_secs(self.0)
+    }
+}
+
+impl TimeFrame {
+    /// Create a new time frame with a specific slot size
+    ///
+    /// ```text
+    ///
+    /// 0        1        2        3        4        5                  
+    /// x--------x--------┳--------x--------x--------x  frame ticking at per_slot
+    ///
+    /// ^
+    /// |
+    /// timeline
+    /// ```
+    ///
+    pub fn new(timeline: Timeline, per_slot: NbSeconds) -> Self {
+        TimeFrame {
+            timeline,
+            slot_offset: Slot(0),
+            slot_duration: per_slot,
+        }
+    }
+
+    /// Change time frame at a specific slot
+    ///
+    /// Note this also change the beginning of this time frame, to start
+    ///
+    /// ```text
+    /// 0        1        2        3        4        5                  
+    /// x--------x--------┳--------x--------x--------x  frame ticking at NbSeconds(9)
+    ///                   |
+    ///                   ┕---x---x---x---x---x         returned frame
+    ///                   2   3   4   5   6   7
+    ///                   ↑
+    ///                   |
+    ///                   frame.change_frame(Slot(2), NbSeconds(4))
+    /// ```
+    ///
+    pub fn change_frame(&self, slot: Slot, duration_per_slot: NbSeconds) -> Self {
+        let d = Duration::from_secs(slot.0 * self.slot_duration.0);
+        let new_timeline = self.timeline.advance(d);
+        TimeFrame {
+            timeline: new_timeline,
+            slot_offset: Slot(self.slot_offset.0 + slot.0),
```

Comment:
Consider making a `Slot::offset` method which would use checked math under the hood.

Reply:
this and the other calculation, I don't really see a point of using checked math at this moment. it does comes with a high prices in term of API for a corner case that we shouldn't really encounter